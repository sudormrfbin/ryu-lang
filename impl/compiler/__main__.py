from compiler.repl import repl

repl()

# with open("input.lang", "r") as f:
#     inp = f.read()

# output = compile(inp)
# print(output)
