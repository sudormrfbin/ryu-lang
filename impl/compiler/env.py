from typing import Any

from .errors import InternalCompilerError


class RuntimeEnvironment:
    values: dict[str, Any]

    def __init__(self):
        self.values = {}

    def define(self, name: str, value: Any):
        # shadowing is allowed
        self.values[name] = value

    def get(self, name: str) -> Any:
        try:
            return self.values[name]
        except KeyError:
            raise InternalCompilerError(f"Variable {name} not defined at runtime")
