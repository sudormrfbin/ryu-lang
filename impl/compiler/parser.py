from typing import Any

from compiler import ast
from compiler.lalr import Token, Tree, Lark_StandAlone, v_args, Transformer, DATA

# https://github.com/lark-parser/lark/issues/565
DATA["options"]["maybe_placeholders"] = True
_parser = Lark_StandAlone(propagate_positions=True)


def parse(source: str) -> Tree[Token]:
    return _parser.parse(source)


def listify(cls):
    @v_args(meta=True, inline=False)
    def _(self, meta, lst):
        return cls(meta, lst)

    return _


@v_args(meta=True, inline=True)
class LarkTreeToAstTransformer(Transformer[Token, Any]):
    def TRUE(self, _) -> bool:
        return True

    def FALSE(self, _) -> bool:
        return False

    def INT(self, n: str) -> int:
        return int(n)

    def STRING(self, s: str) -> str:
        return s[1:-1]  # remove quotes

    statement_list = listify(ast.StatementList)
    statement_block = listify(ast.StatementBlock)
    variable_declaration = ast.VariableDeclaration
    assignment = ast.Assignment
    index_assignment = ast.IndexAssignment
    print_stmt = ast.PrintStmt
    if_chain = ast.IfChain
    if_stmt = ast.IfStmt
    else_if_ladder = listify(ast.ElseIfLadder)
    else_if_stmt = ast.ElseIfStmt
    match_stmt = ast.MatchStmt
    case_ladder = listify(ast.CaseLadder)
    case_stmt = ast.CaseStmt
    enum_pattern = ast.EnumPattern
    enum_pattern_tuple = ast.EnumPatternTuple
    wildcard_pattern = ast.WildcardPattern
    while_stmt = ast.WhileStmt
    for_stmt = ast.ForStmt
    for_stmt_int = ast.ForStmtInt
    struct_stmt = ast.StructStmt
    struct_members = listify(ast.StructMembers)
    struct_member = ast.StructMember
    struct_access = ast.StructAccess
    struct_assignment = ast.StructAssignment
    enum_stmt = ast.EnumStmt
    enum_members = listify(ast.EnumMembers)
    enum_member_bare = ast.EnumMemberBare
    enum_member_tuple = ast.EnumMemberTuple
    function_definition = ast.FunctionDefinition
    function_params = listify(ast.FunctionParams)
    function_param = ast.FunctionParam
    type_annotation = ast.TypeAnnotation
    return_stmt = ast.ReturnStmt
    equality = ast.Equality
    logical = ast.Logical
    comparison = ast.Comparison
    term = ast.Term
    factor = ast.Factor
    unary_op = ast.UnaryOp
    indexing = ast.Indexing
    function_call = ast.FunctionCall
    function_args = listify(ast.FunctionArgs)
    struct_init_members = listify(ast.StructInitMembers)
    struct_init_member = ast.StructInitMember
    bool_literal = ast.BoolLiteral
    int_literal = ast.IntLiteral
    string_literal = ast.StringLiteral
    enum_literal_simple = ast.EnumLiteralSimple
    enum_literal_tuple = ast.EnumLiteralTuple
    variable = ast.Variable
    array_literal = ast.ArrayLiteral
    array_elements = listify(ast.ArrayElements)
    array_element = ast.ArrayElement
    array_pattern = listify(ast.ArrayPattern)
    array_pattern_element = ast.ArrayPatternElement


_transformer = LarkTreeToAstTransformer()


def parse_tree_to_ast(tree: Tree[Token]) -> ast.Ast:
    return _transformer.transform(tree)
