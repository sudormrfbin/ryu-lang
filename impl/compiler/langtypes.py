from __future__ import annotations

from dataclasses import dataclass
from typing import TYPE_CHECKING, Optional, Sequence, TypeAlias
from typing_extensions import override


if TYPE_CHECKING:
    from compiler.env import TypeEnvironment
    from compiler.errors import Span


class Type:
    @property
    def name(self) -> str:
        return type(self).__name__

    @classmethod
    def from_str(cls, ident: str, env: TypeEnvironment) -> Optional["Type"]:
        if ident in PRIMITIVE_TYPES:
            return PRIMITIVE_TYPES[ident]

        ty = env.get(ident)
        if isinstance(ty, UserDefinedType):
            return ty

        return None

    @classmethod
    def from_str_with_generics(
        cls,
        ident: str,
        generics: Type,
        env: TypeEnvironment,
    ) -> Optional["Type"]:
        if ident == "array":
            return Array(generics)

        return None


class Placeholder(Type):
    pass


class PrimitiveType(Type):
    pass


class UserDefinedType(Type):
    pass


class Bool(PrimitiveType):
    pass


class Int(PrimitiveType):
    pass


class String(PrimitiveType):
    pass


class Block(Type):
    pass


@dataclass
class ReturnBlock(Block):
    return_type: Type
    return_stmt_span: Span


@dataclass
class Struct(UserDefinedType):
    struct_name: str
    members: Members

    @dataclass
    class Members(PrimitiveType):
        types: dict[str, Type]

        def __len__(self) -> int:
            return len(self.types)


@dataclass
class Enum(UserDefinedType):
    enum_name: str
    members: list[EnumVariantSimple | EnumVariantTuple]
    span: Span

    @property
    @override
    def name(self) -> str:
        return self.enum_name

    def variant_from_str(
        self, name: str
    ) -> EnumVariantSimple | EnumVariantTuple | None:
        for mem in self.members:
            if mem.name == name:
                return mem
        return None


@dataclass
class EnumVariantSimple:
    name: str


@dataclass
class EnumVariantTuple:
    name: str
    inner: Type


EnumVariants: TypeAlias = EnumVariantSimple | EnumVariantTuple


@dataclass
class Array(PrimitiveType):
    ty: Type

    @property
    @override
    def name(self) -> str:
        return f"Array<{self.ty.name}>"


@dataclass
class UntypedArray(PrimitiveType):
    pass


@dataclass
class Function(PrimitiveType):
    function_name: str
    arguments: Params
    return_type: Type


@dataclass
class Params(PrimitiveType):
    types: list[Type]

    def __len__(self) -> int:
        return len(self.types)


BOOL = Bool()
INT = Int()
STRING = String()
BLOCK = Block()
PLACEHOLDER = Placeholder()


PRIMITIVE_TYPES: dict[str, Type] = {
    "bool": BOOL,
    "int": INT,
    "string": STRING,
}


def resolve_blocks_type(types: Sequence[Type]) -> Block:
    resolved = BLOCK
    for ty in types:
        match (resolved, ty):
            case (Block(), ReturnBlock()):
                resolved = ty
            case (ReturnBlock(), ReturnBlock()):
                if resolved.return_type != ty.return_type:
                    raise  # TODO different return types
            case _:
                pass

    return resolved
