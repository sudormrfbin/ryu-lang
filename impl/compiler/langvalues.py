from abc import abstractmethod
from dataclasses import dataclass
from typing import TYPE_CHECKING, Any
from typing_extensions import override
from compiler import runtime

from compiler.env import RuntimeEnvironment

if TYPE_CHECKING:
    from compiler.ast import StatementBlock


@dataclass
class EnumValue:
    ty: str
    variant: str

    @override
    def __str__(self) -> str:
        return f"{self.ty}::{self.variant}"

    @override
    def __hash__(self) -> int:
        return hash(repr(self))


class Function:
    @abstractmethod
    def call(self, args: list[Any], env: "RuntimeEnvironment") -> Any:
        pass


@dataclass
class RyuFunction(Function):
    param_names: list[str]
    body: "StatementBlock"

    @override
    def call(self, args: list[Any], env: "RuntimeEnvironment") -> Any:
        child_env = RuntimeEnvironment(enclosing=env)

        for name, arg in zip(self.param_names, args):
            child_env.define(name, arg)

        try:
            self.body.eval(child_env)
        except runtime.FunctionReturn as ret:
            return ret.return_value
