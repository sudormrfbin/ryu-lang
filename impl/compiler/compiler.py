from typing import Any

from compiler.env import RuntimeEnvironment, TypeEnvironment

from compiler import builtins, errors, langvalues
from compiler.parser import parse, parse_tree_to_ast

BUILTIN_FUNCTIONS: list[type[langvalues.BuiltinFunction]] = [
    builtins.SumFunction,
    builtins.ArrayLengthFunction,
    builtins.StringLengthFunction,
    builtins.ArrayAppend,
]


def get_default_environs() -> tuple[TypeEnvironment, RuntimeEnvironment]:
    type_env = TypeEnvironment()
    runtime_env = RuntimeEnvironment()

    for fn in BUILTIN_FUNCTIONS:
        type_env.define(fn.TYPE.function_name, fn.TYPE)
        runtime_env.define(fn.TYPE.function_name, fn())

    return (type_env, runtime_env)


def run(source: str, type_env: TypeEnvironment, runtime_env: RuntimeEnvironment) -> Any:
    try:
        return _run(source, type_env, runtime_env)
    except errors.CompilerError as err:
        err.report(source)


def _run(
    source: str, type_env: TypeEnvironment, runtime_env: RuntimeEnvironment
) -> Any:
    tree = parse(source)
    ast = parse_tree_to_ast(tree)

    ast.typecheck(type_env)

    return ast.eval(runtime_env)
