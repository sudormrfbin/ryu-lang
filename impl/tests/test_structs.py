from typing import Any
from compiler.env import RuntimeEnvironment, TypeEnvironment
from compiler.langvalues import StructValue
from compiler.parser import parse, parse_tree_to_ast
from compiler.langtypes import INT, STRING, Members, Struct
from tests.utils import docstring_source_with_snapshot


# 1. struct type in typeenv


@docstring_source_with_snapshot
def test_struct_def(source: str, snapshot: Any):
    """
    struct Person {
        name: string
        age: int
    }
    """
    ast = parse_tree_to_ast(parse(source))
    assert ast.to_dict() == snapshot

    type_env = TypeEnvironment()
    ast.typecheck(type_env)
    assert ast.to_type_dict() == snapshot

    assert type_env.get("Person") == Struct(
        struct_name="Person",
        members=Members({"name": STRING, "age": INT}),
    )


@docstring_source_with_snapshot
def test_struct_init(source: str, snapshot: Any):
    """
    struct Person {
        name: string
        age: int
    }

    let p = Person(name="bob", age=23)
    """
    ast = parse_tree_to_ast(parse(source))
    assert ast.to_dict() == snapshot

    type_env = TypeEnvironment()
    ast.typecheck(type_env)
    assert ast.to_type_dict() == snapshot

    env = RuntimeEnvironment()
    ast.eval(env)
    assert env.get("p") == StructValue(
        name="Person",
        attrs={
            "name": "bob",
            "age": 23,
        },
    )
