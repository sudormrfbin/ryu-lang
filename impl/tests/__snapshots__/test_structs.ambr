# serializer version: 1
# name: test_struct_def
  dict({
    <class 'compiler.ast.StructStmt'>: dict({
      'members': dict({
        <class 'compiler.ast.StructMembers'>: dict({
          'members': list([
            dict({
              <class 'compiler.ast.StructMember'>: dict({
                'ident_type': dict({
                  <class 'compiler.ast.TypeAnnotation'>: dict({
                    'ty': Token('IDENTIFIER', 'string'),
                  }),
                }),
                'name': Token('IDENTIFIER', 'name'),
              }),
            }),
            dict({
              <class 'compiler.ast.StructMember'>: dict({
                'ident_type': dict({
                  <class 'compiler.ast.TypeAnnotation'>: dict({
                    'ty': Token('IDENTIFIER', 'int'),
                  }),
                }),
                'name': Token('IDENTIFIER', 'age'),
              }),
            }),
          ]),
        }),
      }),
      'name': Token('IDENTIFIER', 'Person'),
    }),
  })
# ---
# name: test_struct_def.1
  dict({
    <class 'compiler.ast.StructStmt'>: dict({
      'fields': dict({
        'members': dict({
          <class 'compiler.ast.StructMembers'>: dict({
            'fields': dict({
              'members': list([
                dict({
                  <class 'compiler.ast.StructMember'>: dict({
                    'fields': dict({
                      'ident_type': dict({
                        <class 'compiler.ast.TypeAnnotation'>: dict({
                          'fields': dict({
                          }),
                          'type': <class 'compiler.langtypes.String'>,
                        }),
                      }),
                    }),
                    'type': <class 'compiler.langtypes.String'>,
                  }),
                }),
                dict({
                  <class 'compiler.ast.StructMember'>: dict({
                    'fields': dict({
                      'ident_type': dict({
                        <class 'compiler.ast.TypeAnnotation'>: dict({
                          'fields': dict({
                          }),
                          'type': <class 'compiler.langtypes.Int'>,
                        }),
                      }),
                    }),
                    'type': <class 'compiler.langtypes.Int'>,
                  }),
                }),
              ]),
            }),
            'type': <class 'compiler.langtypes.Struct.Members'>,
          }),
        }),
      }),
    }),
  })
# ---
# name: test_struct_init
  dict({
    <class 'compiler.ast.StatementList'>: dict({
      'stmts': list([
        dict({
          <class 'compiler.ast.StructStmt'>: dict({
            'members': dict({
              <class 'compiler.ast.StructMembers'>: dict({
                'members': list([
                  dict({
                    <class 'compiler.ast.StructMember'>: dict({
                      'ident_type': dict({
                        <class 'compiler.ast.TypeAnnotation'>: dict({
                          'ty': Token('IDENTIFIER', 'string'),
                        }),
                      }),
                      'name': Token('IDENTIFIER', 'name'),
                    }),
                  }),
                  dict({
                    <class 'compiler.ast.StructMember'>: dict({
                      'ident_type': dict({
                        <class 'compiler.ast.TypeAnnotation'>: dict({
                          'ty': Token('IDENTIFIER', 'int'),
                        }),
                      }),
                      'name': Token('IDENTIFIER', 'age'),
                    }),
                  }),
                ]),
              }),
            }),
            'name': Token('IDENTIFIER', 'Person'),
          }),
        }),
        dict({
          <class 'compiler.ast.VariableDeclaration'>: dict({
            'ident': Token('IDENTIFIER', 'p'),
            'rvalue': dict({
              <class 'compiler.ast.FunctionCall'>: dict({
                'args': dict({
                  <class 'compiler.ast.StructInitMembers'>: dict({
                    'members': list([
                      dict({
                        <class 'compiler.ast.StructInitMember'>: dict({
                          'name': Token('IDENTIFIER', 'name'),
                          'value': dict({
                            <class 'compiler.ast.literals.StringLiteral'>: dict({
                              'value': 'bob',
                            }),
                          }),
                        }),
                      }),
                      dict({
                        <class 'compiler.ast.StructInitMember'>: dict({
                          'name': Token('IDENTIFIER', 'age'),
                          'value': dict({
                            <class 'compiler.ast.literals.IntLiteral'>: dict({
                              'value': 23,
                            }),
                          }),
                        }),
                      }),
                    ]),
                  }),
                }),
                'callee': dict({
                  <class 'compiler.ast.Variable'>: dict({
                    'value': Token('IDENTIFIER', 'Person'),
                  }),
                }),
              }),
            }),
          }),
        }),
      ]),
    }),
  })
# ---
# name: test_struct_init.1
  dict({
    <class 'compiler.ast.StatementList'>: dict({
      'fields': dict({
        'stmts': list([
          dict({
            <class 'compiler.ast.StructStmt'>: dict({
              'fields': dict({
                'members': dict({
                  <class 'compiler.ast.StructMembers'>: dict({
                    'fields': dict({
                      'members': list([
                        dict({
                          <class 'compiler.ast.StructMember'>: dict({
                            'fields': dict({
                              'ident_type': dict({
                                <class 'compiler.ast.TypeAnnotation'>: dict({
                                  'fields': dict({
                                  }),
                                  'type': <class 'compiler.langtypes.String'>,
                                }),
                              }),
                            }),
                            'type': <class 'compiler.langtypes.String'>,
                          }),
                        }),
                        dict({
                          <class 'compiler.ast.StructMember'>: dict({
                            'fields': dict({
                              'ident_type': dict({
                                <class 'compiler.ast.TypeAnnotation'>: dict({
                                  'fields': dict({
                                  }),
                                  'type': <class 'compiler.langtypes.Int'>,
                                }),
                              }),
                            }),
                            'type': <class 'compiler.langtypes.Int'>,
                          }),
                        }),
                      ]),
                    }),
                    'type': <class 'compiler.langtypes.Struct.Members'>,
                  }),
                }),
              }),
            }),
          }),
          dict({
            <class 'compiler.ast.VariableDeclaration'>: dict({
              'fields': dict({
                'rvalue': dict({
                  <class 'compiler.ast.FunctionCall'>: dict({
                    'fields': dict({
                      'args': dict({
                        <class 'compiler.ast.StructInitMembers'>: dict({
                          'fields': dict({
                            'members': list([
                              dict({
                                <class 'compiler.ast.StructInitMember'>: dict({
                                  'fields': dict({
                                    'value': dict({
                                      <class 'compiler.ast.literals.StringLiteral'>: dict({
                                        'fields': dict({
                                        }),
                                        'type': <class 'compiler.langtypes.String'>,
                                      }),
                                    }),
                                  }),
                                  'type': <class 'compiler.langtypes.String'>,
                                }),
                              }),
                              dict({
                                <class 'compiler.ast.StructInitMember'>: dict({
                                  'fields': dict({
                                    'value': dict({
                                      <class 'compiler.ast.literals.IntLiteral'>: dict({
                                        'fields': dict({
                                        }),
                                        'type': <class 'compiler.langtypes.Int'>,
                                      }),
                                    }),
                                  }),
                                  'type': <class 'compiler.langtypes.Int'>,
                                }),
                              }),
                            ]),
                          }),
                          'type': <class 'compiler.langtypes.Struct.Members'>,
                        }),
                      }),
                      'callee': dict({
                        <class 'compiler.ast.Variable'>: dict({
                          'fields': dict({
                          }),
                          'type': <class 'compiler.langtypes.Struct'>,
                        }),
                      }),
                    }),
                    'type': <class 'compiler.langtypes.Struct'>,
                  }),
                }),
              }),
            }),
          }),
        ]),
      }),
    }),
  })
# ---
